Document::ListenerType should be an enum class

rniwa opened the pull request:
#### b4ade8e6bda5fd5d1d936c513d4fe88d73e88287
<pre>
Document::ListenerType should be an enum class
<a href="https://bugs.webkit.org/show_bug.cgi?id=256492">https://bugs.webkit.org/show_bug.cgi?id=256492</a>

Reviewed by Chris Dumez.

Made Document::ListenerType an enum class and made m_listenerTypes an OptionSet.

Also introduced a new helper function: hasListenerTypeForEventType.

* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::shouldFireDOMEvents const):
* Source/WebCore/dom/CharacterData.cpp:
(WebCore::canUseSetDataOptimization):
(WebCore::CharacterData::dispatchModifiedEvent):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::dispatchChildInsertionEvents):
(WebCore::dispatchChildRemovalEvents):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasListenerTypeForEventType const):
(WebCore::Document::addListenerTypeIfNeeded):
* Source/WebCore/dom/Document.h:
(WebCore::Document::hasListenerType const):
(WebCore::Document::hasAnyListenerOfType const): Added.
(WebCore::Document::addListenerType):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchFocusInEventIfNeeded):
(WebCore::Element::dispatchFocusOutEventIfNeeded):
(WebCore::Element::dispatchMouseForceWillBegin):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dispatchSubtreeModifiedEvent):
* Source/WebCore/editing/markup.cpp:
(WebCore::hasMutationEventListeners):
* Source/WebCore/history/CachedFrame.cpp:
(WebCore::CachedFrameBase::restore):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::OverflowEventDispatcher::OverflowEventDispatcher):

Canonical link: <a href="https://commits.webkit.org/263844@main">https://commits.webkit.org/263844@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/832b11292f0e8a4b8e5572343426ff5d2d7d6acc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6041 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7423 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/6241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5869 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6003 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/8393 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6017 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5316 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7482 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3494 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5291 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/5360 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5370 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/7584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5819 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4772 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9377 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/690 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5615 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Document.h:
```diff
@@ -940,27 +940,26 @@ class Document
-        DOMNODEINSERTEDINTODOCUMENT_LISTENER = 1 << 4,
-        DOMCHARACTERDATAMODIFIED_LISTENER    = 1 << 5,
-        OVERFLOWCHANGED_LISTENER             = 1 << 6,
-        TRANSITIONEND_LISTENER               = 1 << 7,
```
This is never used so I've gotten rid of it.

cdumez commented on Source/WebCore/dom/Document.h:
```diff
@@ -2243,7 +2242,7 @@ class Document
     unsigned m_dataListElementCount { 0 };
 
-    unsigned m_listenerTypes { 0 };
+    OptionSet<ListenerType> m_listenerTypes { };
```
We probably don't need the `{ }`

rniwa replied:
Good point.

cdumez commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -1458,11 +1458,11 @@ static inline bool hasOneTextChild(ContainerNode& node)
-        || document.hasListenerType(Document::DOMNODEREMOVED_LISTENER)
-        || document.hasListenerType(Document::DOMNODEREMOVEDFROMDOCUMENT_LISTENER)
-        || document.hasListenerType(Document::DOMCHARACTERDATAMODIFIED_LISTENER);
+    return document.hasListenerType(Document::ListenerType::DOMSubtreeModified)
```
Now that we use an OptionSet, it may be nice to leverage OptionSet::containsAny(). Maybe it would look something like:
```
return document.hasAnyListenerOfType({ Document::ListenerType::DOMSubtreeModified, ListenerType::DOMNodeInserted, ... });
```

Just a suggestion. You don't have to do it in this patch.

cdumez reviewed (approved):
LGTM

rniwa committed:
Document::ListenerType should be an enum class
https://bugs.webkit.org/show_bug.cgi?id=256492

Reviewed by Chris Dumez.

Made Document::ListenerType an enum class and made m_listenerTypes an OptionSet.

Also introduced a new helper function: hasListenerTypeForEventType.

* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::shouldFireDOMEvents const):
* Source/WebCore/dom/CharacterData.cpp:
(WebCore::canUseSetDataOptimization):
(WebCore::CharacterData::dispatchModifiedEvent):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::dispatchChildInsertionEvents):
(WebCore::dispatchChildRemovalEvents):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasListenerTypeForEventType const):
(WebCore::Document::addListenerTypeIfNeeded):
* Source/WebCore/dom/Document.h:
(WebCore::Document::hasListenerType const):
(WebCore::Document::hasAnyListenerOfType const): Added.
(WebCore::Document::addListenerType):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchFocusInEventIfNeeded):
(WebCore::Element::dispatchFocusOutEventIfNeeded):
(WebCore::Element::dispatchMouseForceWillBegin):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dispatchSubtreeModifiedEvent):
* Source/WebCore/editing/markup.cpp:
(WebCore::hasMutationEventListeners):
* Source/WebCore/history/CachedFrame.cpp:
(WebCore::CachedFrameBase::restore):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::OverflowEventDispatcher::OverflowEventDispatcher):

Canonical link: https://commits.webkit.org/263844@main
